src/scripts/codes/RedeemableCodes.ts

```diff
@@ -44,70 +44,25 @@ class RedeemableCodes implements Saveable {
-                GameConstants.JohtoDungeons.forEach(dungeon => {
-                    GameHelper.incrementObservable(App.game.statistics.dungeonsCleared[Statistics.getDungeonIndex(dungeon)]);
-                });
-                // Catch all Pokemon
-                for (let id = GameConstants.TotalPokemonsPerRegion[GameConstants.Region.kanto] + 1; id <= GameConstants.TotalPokemonsPerRegion[GameConstants.Region.johto]; id++) {
-                    App.game.party.gainPokemonById(id, false, true);
-                }
-                // Notify that the code was activated successfully
-                Notifier.notify({ title:'Code activated!', message: 'You have unlocked all of the Johto region', type: GameConstants.NotificationOption.success, timeout: 1e4 });
-            }),
-            new RedeemableCode('complete-hoenn', -1257697040, false, function () {
-                // Complete all routes
-                for (let route = GameConstants.RegionRoute[GameConstants.Region.hoenn][0]; route <= GameConstants.RegionRoute[GameConstants.Region.hoenn][1]; route++) {
-                    GameHelper.incrementObservable(App.game.statistics.routeKills[route], 10);
-                }
-                // Complete all gyms
-                GameConstants.HoennGyms.forEach(gym => {
-                    GameHelper.incrementObservable(App.game.statistics.gymsDefeated[Statistics.getGymIndex(gym)]);
-                    // Give badge
-                    if (!App.game.badgeCase.hasBadge(gymList[gym].badgeReward)) {
-                        App.game.badgeCase.gainBadge(gymList[gym].badgeReward);
-                    }
-                });
-                // Complete all dungeons
-                GameConstants.HoennDungeons.forEach(dungeon => {
-                    GameHelper.incrementObservable(App.game.statistics.dungeonsCleared[Statistics.getDungeonIndex(dungeon)]);
-                });
-                // Catch all Pokemon
-                for (let id = GameConstants.TotalPokemonsPerRegion[GameConstants.Region.johto] + 1; id <= GameConstants.TotalPokemonsPerRegion[GameConstants.Region.hoenn]; id++) {
-                    App.game.party.gainPokemonById(id, false, true);
-                }
-                // Notify that the code was activated successfully
-                Notifier.notify({ title:'Code activated!', message: 'You have unlocked all of the Hoenn region', type: GameConstants.NotificationOption.success, timeout: 1e4 });
-            }),
         ];
     }
 
     enterCode(code: string) {
         const hash = this.hash(code);
 
-        const redeemableCode = this.codeList.find(code => {
-            return code.hash === hash;
+        const redeemableCode = this.codeList.find(c => {
+            return c.hash === hash;
         });
 
-        if (!redeemableCode) {
-            Notifier.notify({ message: `Invalid code ${code}`, type: GameConstants.NotificationOption.danger });
+        if (!redeemableCode && !App.game.discord.checkCode(code)) {
```

Comment:
I'm not so sure about this. It is entirely possible that someone won't be able to redeem a discord code because its hash collides with one of the RedeemableCodes. I think you need some prefix before the code we can check for instead.

We could also do with some kind of `Redeemable` interface that `RedeemableCode`s and `DiscordCode`s implement, and just have App.game.discord return the redeemable thing for this function to work with. I think that would make this function a little clear, just going by its name `checkCode` doesn't seem like something that would do the actual redeeming.

Reply:
> I'm not so sure about this. It is entirely possible that someone won't be able to redeem a discord code because its hash collides with one of the RedeemableCodes. I think you need some prefix before the code we can check for instead.

The Discord codes all currently have `XXXX-XXXX-XXXX` so i've added a regex check.

> We could also do with some kind of Redeemable interface that RedeemableCodes and DiscordCodes implement, and just have App.game.discord return the redeemable thing for this function to work with. I think that would make this function a little clear, just going by its name checkCode doesn't seem like something that would do the actual redeeming.

I've update the function name to `enterCode` it's still not the best, I'm just not entirely sure how to clean it up better.
I originally had all the Discord codes just in the `RedeemableCodes` area, but quickly realised that wasn't working very well 😅.